### ISSUE TITLE
Request: Proper usage of Task return in scoped statements

### ISSUE THREAD

**Issue by OlegKarasik** [JosefPihrt/Roslynator#419]
The idea originates from this [tweet](https://twitter.com/marcgravell/status/1023176337652109312).

Consider the following example:

``` csharp
// Pay attention that method isn't marked as **async**
Task Method()
{
  using (resource)
  {
    // .. code executed inside using
    return CallAsync();
  }
}
```

Here we can see one of two situations:
1. The `CallAsync` has to be executed inside the `using` statement.
2. The `CallAsync` has nothing related to `using` statement and was placed here as part of the logic flow.

The analyzer would suggest two things:
1. Mark method as `async` and use `await` inside the `lock` statement:
  ``` csharp
  async Task Method()
  {
    using (resource)
    {
      // .. code executed inside using
      return await CallAsync();
    }
  }
  ```
2. Transform code to explicitly show that `CallAsync` isn't part of the `using` and isn't related to `resource`:
  ``` csharp
  Task Method()
  {
    Task t;
    using (resource)
    {
      // .. code executed inside using
      t = CallAsync();
    }
    return t;
  }
  ```

**UPDATE 7/30/2018**
Updated code sample to use `using` instead of `lock` because as pointed by @mgravell the usage of `lock` has invalid semantics here.

**Comment by mgravell** [comment]
Note: 1. isn't a valid code change; you can't `await` inside a `lock` due to the threading semantics - it needs to switch to a different code implementation (probably `SemaphoreSlim`). That change *can* be done for `using`, however.

**Comment by OlegKarasik** [comment]
@mgravell 
Yes you are absolutely right. I have modified the code samples to use `using` instead of `lock`.

**Comment by JosefPihrt** [comment]
Sounds like a good idea.

**Comment by JosefPihrt** [comment]
fixed with 08990abc4806bb4de1172ac382314f1aec9ecf08
